### ISSUE TITLE
How fast should re-replication be initiated?

### ISSUE THREAD

**Issue by eriklindahl** [lizardfs/lizardfs#589]
Hi,

Before putting our filesystem to real use, we figured we should stress-test it a bit. We have nice automated failover (and recovery) of the master working with keepalived, and as another test we want to see what happens if a disk just disappears.

We don't have any real precious data on it yet, but it's a live-size installation with 7 servers and some 20 million chunks over 75 disks. For now we're using 4+1 erasure coding to test the EC code, since the plan is to later move to 8+2 with more servers.

We are using separate ZFS pools for each disk, with a lizardfs directory at the top so we can detect that the path is there.

After doing "rm -rf *" on one of these disks with ~500GB of data, I correctly see errors both on the chunkserver and master. All files are still readable, so things are looking good there. The cgi-admin page also shows there are errors on this disk (in the last-error column, the status still says "ok" for the disk). 

However, even after a few hours, master still lists all chunks as stable, and I haven't seen anything about replication taking place. I would assume I should see *something* about it in the background or logs. Of course I'm well aware just removing files is a pretty nasty error (no hard device errors, etc.), but from the LizardFS p-o-v the only thing that should matter is that the files cannot be found...

So, what is the process where the chunk/master server moves from having detected a last-error to actually triggering re-replication?

**Comment by psarna** [comment]
Reporting a faulty chunk is done by the chunkserver that owns it. There are two ways of discovering a faulty chunk - first one is periodic chunk test preformed continuously by chunkservers (by default it's something like 1 chunk per 10 seconds, configurable through `mfschunkserver.cfg`). The second way is trying to perform an I/O operation on such chunk. If read/write to a chunk fails then it's marked as damaged.
After that, master server immediately receives an information that this particular chunk is damaged.
Then, everything happens within the master server, which periodically processes every chunk and checks if any action is needed. One of the action might be replication, if there are not enough copies present in the system. This periodical check is called 'chunk loop' and it's configurable via `mfsmaster.cfg`.

You can browse through `man mfsmaster.cfg` and `man mfschunkserver.cfg` to find various options - how often replications occur, what's the replication delay after a chunkserver disconnects, how much CPU should 'chunk loop' take, etc.

**Comment by eriklindahl** [comment]
Hi Piotr,

Not a single chunk was marked as undergoal in 5h (despite forcibly removing 500k chunks, and this is only one of three disks on the server), so there seems to be some part of the automated tests that isn't working as intended. 

Any idea how to be debug it?

**Comment by psarna** [comment]
And how do you check if a chunk was marked undergoal? I'm asking because CGI is an overview of the system, but it's not real-time up-to-date. There are "check loop start time" and "check loop end time" entries in Info tab that show the last time chunk information was gathered and updated.

It's way more reliable to pick a file, remove some of its chunks for testing purposes and then check it with `lizardfs fileinfo FILE_PATH`. That way you can see LizardFS has acknowledged that this chunk is damaged. It may take your chunkserver some time to auto-discover it (especially if there are a lot of chunks there), so you can try reading the file a couple of times in order to discover chunk rot (and read with O_DIRECT, otherwise you may end up reading from your own cache of course).

**Comment by eriklindahl** [comment]
If the replication had worked, I assume these chunks would not suddenly go from stable to undergoal and replication started when the chunkserver process was restarted and the disk scan initiated, right?

**Comment by eriklindahl** [comment]
PS: In particular for long-term archival purposes, it is of course critical that the server doesn't postpone checking the chunk health until we read a file, so that's why I don't want to try and force it :-)

**Comment by guestisp** [comment]
> PS: In particular for long-term archival purposes, it is of course critical that the server doesn't postpone checking the chunk health until we read a file, so that's why I don't want to try and force it :-)

I don't know Lizard very well but there is a process that scan all chunkserver looking for chunks. If anything bad is found, replication is started. This process is low-priority and doesn't run often and doesn't scan the whole chunkserver at once. Some time is needed. (I think this is tunable)

**Comment by guestisp** [comment]
> you can try reading the file a couple of times in order to discover chunk rot 

Why "a couple of time" and not only one? 
Are you saying that reading a file only once, could not trigger immediatly a replication?

**Comment by psarna** [comment]
Yes, that just means that your chunkserver hasn't discovered these broken chunks yet. Restarting it triggered a scan of all files. You can change the value of
~~~
       HDD_TEST_FREQ
           chunk test period in seconds (default is 10)
~~~
from mfschunkserver.cfg to something lower if you want to see the results faster. Anyway, it may take a while, especially when there are a lot of chunks on your chunkserver. The reason why this test doesn't happen very often is performance - extensive scanning would cause constant high I/O on your chunkservers.

Every damaged chunk will be discovered eventually.

**Comment by psarna** [comment]
@guestisp
I meant **potentially** couple of times if chunk has more than 1 copy. Reading more times makes it more probable that every copy would be read.

**Comment by eriklindahl** [comment]
I just realized one potential reason - is the HDD test performed serially for one drive at a time? I kind of presumed it would run in parallel on all drives, and then I would expect to see at least one error 5h after having removed *all* chunks on the drive?

**Comment by guestisp** [comment]
how many chunks are scanned every time ? I dubt that in 10 seconds (default of `HDD_TEST_FREQ`) you'll scan all chunks, thus, every 10 seconds you should scan only a portion. Which is the size of this portion ?

In an archive host, where files are not read often, you could stay with undergoal chunks too much.

**Comment by eriklindahl** [comment]
I can't find any information about scanned chunks, but I'll try to increase logging levels and redo it tonight. Just need to let the rebuild finish :-)   If it actually was scanning that hard drive, it would be sufficient for it to check a single chunk in 5h, since they are all gone.

**Comment by 4Dolio** [comment]
I think the problem is that your test does not match a practical failure that would be encountered.  If you pulled the physical disk from the server then you will get a different result perhaps..  I agree that perhaps there is some room for improvement.  The chunk check loop is not what you are looking for, should only be expected to identify rare instances of disk/chunk rot.  Even if you turned up that check it would take months for all chunks to be checked.  What would be useful is that all your disks should show now errors ever.. As soon as one has an last error on, and in your test this should happen such that the last error will keep changing and always be "recent".  Seeing such errors or any errors on any disk should prompt a human to look into it more closely.  You would discover that a huge number of your chunks were missing.. or you would restart the chunk server, inducing a full scan of all chunks, immediately revealing to the master that 500GB worth of chunks for that disk which the master knows that CS should have are now missing.  Master records them as missing and replicates to correct the missing chunks.

**Comment by eriklindahl** [comment]
Since LizardFS doesn't work with the raw device, I'm not sure if pulling a disk would show up in any other way than the entire mount poing being gone, but it's a good idea to test it. However, my experience with failing disks is that it typically starts showing up as individual sectors that cannot be read, and eventually no files can be read. And, while these failures would be rare, a complete corruption of the filesystem where the data is stored is definitely a possibility.  I do see the challenges with automatically running scans, but this seems like a fairly large risk of almost-silent data loss - unless the sysadmin keeps checking for "last error" reports and then initiates a manual disk check.  Under what conditions apart from a full manual scan will LizardFS mark a storage path as "damaged"?

Wouldn't the obvious solution (which I kind of expected was already used) be for a chunkserver to initiate a scan on a specific disk when errors are reported? If there is any way to access this status without going through the web interface I can even automated it with scripts, although it would be nice to limit the scan to the disk showing errors.

**Comment by 4Dolio** [comment]
Humm.. Since it is not a raw disk, unmounting or otherwise offlining the storage should be comparable and result in a disk fault detection.  Otherwise redundancy is what keeps the data safe until after a very long time without intervention, all chunks are identified as missing...

**Comment by 4Dolio** [comment]
That is an interesting idea for a new feature.  If X(tunable) errors are encountered in X(tunable) timespan then consider the disk faulty...

**Comment by guestisp** [comment]
> That is an interesting idea for a new feature. If X(tunable) errors are encountered in X(tunable) timespan then consider the disk faulty... I believe there is a "last error" and an "error count" for each disk already.

SMART does this automatically, by placing disk in a predictive failure when X errors are encountered in X timespan.

**Comment by 4Dolio** [comment]
Yes, and it would be nice if LFS would convey SMART up to the CS, but for reasons beyond this topic it is too complex to be a core feature at this time.  But you should probably be paying attention to SMART via other methods available.

**Comment by 4Dolio** [comment]
Tuning all comes down to personal preferences... I intentionally use a high replication delay such that if a disk/server were to fail It does not auto replicate for many hours.  The changes are I can recover the service and nothing is permanently lost, and I run the cluster near enough their capacity that if they were to replicate to maintain goals they would exhaust all space.  I do not want that to occur without intervention, so If I did in fact lose the data I would need to modify goals and then let it replicate back to the new baseline.  So while it may seem like 'Why did it not fix itself' is a good question, there may good reasons things do not work as you might at face value expect them to..  It does need some oversight, I try to give the cgi a glance at least once a day.  But we have had hours or days with a 'major' fault of disk/CS but no lose due to the inherent redundancy of the cluster.  It would be difficult to properly automatically handle all fault scenarios..

**Comment by eriklindahl** [comment]
I'm all for configurability when it comes to action. However, I think the basic question is simpler: given that LFS uses paths in the filesystem for storage, I think it is reasonable to expect LFS to notice if the storage location is entirely gone and all data is missing, rather than classify it as "OK". Based on the documentation, I don't think the average user is aware they risk silent data loss unless they manually check disks daily :-)

**Comment by psarna** [comment]
@guestisp 
The portion size for chunk test is 1.

@eriklindahl 
Generally the point of periodic chunk tests is to detect rot, but don't read to much in order to minimize the performance impact. When any chunk test fails, it is properly logged in syslog (and "Important messages" section in CGI), so the proper action after seeing such a thing is to manually ensure that everything is fine with the disk (restarting the chunkserver and trigerring path scanning is good enough a check).
So, what we can do is add more warnings and perhaps make it more vivid in CGI that a chunk test failed.
"Storage location is entirely gone" is rather a specific case and there is no special handling of it (or a special message). What usually happens is that device switches to read-only mode, which still leaves chunks readable. If it solves your issue, we can just add additional warnings somewhere in our CGI that a chunk file was not present during the test, so there might be something wrong with the whole path.

**Comment by eriklindahl** [comment]
Hi Piotr,

First let me re-emphasize what an amazing piece of software I find LFS to be. This is not fundamental critique of the project :-)

However, I hope you agree at least in part that what you write above about checking logs and manually check disks is *completely* different from the statement on the first page in the lizardfs white paper:

"Disk and server failures are handled transparently, without any downtime"

**Comment by guestisp** [comment]
> The portion size for chunk test is 1.

Are you checking 1 chunk every `HDD_TEST_FREQ` seconds ?

**Comment by eriklindahl** [comment]
PS: To look forward to solutions, based on the considerations in this thread it seems like it would work well to detect when these disk IO errors happen, and if they pass a certain threshold (which can be set high) we issue a check of just that disk. 

1) Is there presently any way to use a command-line tool to get diagnostics about what disks have reported errors recently, e.g. like the last-error time from the web interface?

2) Is there any way to force a check of just a single disk from the command line?

**Comment by 4Dolio** [comment]
I was mistaken and there is not an error_count in the cgi similar to last_error.  Maybe there already is an internal counter, I suspect not.  If there were one then a tunable for error_count could trigger a re-scan disk.  I also do not believe there is a "re-scan disk(s)" function presently.  You could emulate having one by removing a disk from the hdd.cfg (or mark a disk for removal without removing it completely), reloading CS, add the disk back, reload CS.  I think that should work, maybe the mark for removal would not trigger a re-scan once no longer marked for removal?

**Comment by 4Dolio** [comment]
Speaking frankly, manually deleting chunk files should never be encountered naturally.  So IMHO that is not really a valid test as it would never happen.  You ext4 or zfs might become corrupt, encounter I/O Errors, Go Read Only, etc and LFS should definitely flag the disk as broken when those happen, and I believe that it already does.  Maybe, if the parent path for a hdd.cfg were to go missing then CS should detect ( unable to cd to parent folder, nor 00..FF sub-folders, and trigger a disk fault.  But if you just deleted the chunk files from within the 00..FF folders, and not those folders, nor their parent, then I do not expect LFS to handle that, that should never happen, if it does then something else is to blame...

**Comment by eriklindahl** [comment]
Update/conclusion:

It's possible to see disk errors through the lizardfs-admin tool, so it's possible to write a small script to check that. However, I also decided to test and just offline/delete a device. It might be useful to know that the default ZFS config will just hang all IO if an entire (single-disk) pool fails, which in turn will cause the entire LFS filesystem to hang (despite server-level redundancy).

When using XFS these problems never appear. Setting the ZFS failmode option to "continue" seems to to the trick there too.

Just to continue to create nasty errors, I decided to zero-fill part of a device :-) ZFS finds this on the next scrub, and reports missing data, but since the filesystem is still "online", LFS is not aware of the errors.

With larger disks we might eventually have up to 100M chunks/drive, so I realize it might take a VERY long time for LFS' own bit-rot scanning to find this.  Since we anyway have redundancy on the server level, my plan is that we simply create a cron job to restart the chunkserver process in the middle of the night once a month or so, and make sure we use different days for different servers. This should force the LFS scan and detect any silent errors. Since the system will remain online, I'm perfectly fine with potentially having slightly lower throughput for a few minutes at 4am.

 Thanks for all the help & recommendations!

**Comment by guestisp** [comment]
@4Dolio 
> But if you just deleted the chunk files from within the 00..FF folders, and not those folders, nor their parent, then I do not expect LFS to handle that, that should never happen, if it does then something else is to blame...

not true. I had multiple crashed with XFS or EXT4. The result was a corrupted FS where some files was missing, other that this, disk was perfectly fine.

There are multiple cases where you could loose some files in a directory and not the whole directory/fs.
(ie: a bad developed script, some FS crashes and so on)

**Comment by guestisp** [comment]
@eriklindahl 
ZFS will detect a "broken file" on every read, not only on a scrub. If you "zero-fill" part of a device and you read a file with part in that zone, ZFS will detect that thank to it's bit-rot protection (the file checksum won't match)

Instead of restarting a chunkserver every night, would be better to add a manual scrub that walks the whole storage. I know that LFS is able to scrub chunks, but it does in a very very very slow way.
A manual scrub that traverse the whole storage would be great, just to be sure that all replicas are there.

All RAID (hw or sw) and even ZFS has scrub features to prevent data-loss. LFS lacks this and IMHO should be added.

**Comment by 4Dolio** [comment]
The handling of unusual and even predictable disk and thus chunk faults could always be improved, deeper detection, better feedback, etc... However, I do not have many real world examples of a Chunk Server going sideways given that you know what to expect.  It sounds like we have made progress in that regards, malicious experimentation is a good means to that end.

I have at times gotten excessively hands, mostly to learn more, with chunk file data outside the purview and control of the CS services abilities.  Such as doing things like background overnight rebalancing of numerous disks within a jbod.  I still use some hardware RAID5/6 to back LFS hdd.cfg storage but have slowly come to prefer simply throwing a raw disk (pick your favorite FS flavor) at the CS and allow LFS to do it's job and recover from faults, enhance scaling, and maintain constant availability.  Raw disk use let's you combine tools and do things like keep a smart failing disk in service for some months longer than you should, while having a hot standby rsync-ed replacement which will only loose a short timespan worth of data from outdated chunks while most chunks remain valid, resulting in just a few minutes to fully recover versus a perfectly acceptable hours or days were LFS left to recover naturally.

> There are multiple cases where you could loose some files in a directory and not the whole directory/fs. (ie: a bad developed script, some FS crashes and so on)

Of course.. I presume too much perhaps.  Such as your CS is secure from any direct user interactions to anyone less privileged than unix wheel members.  There is highly restricted physical access preventing accidental removal of a live drive from a chassis.  I have not yet, but have considered investigating leveraging chunk disks on zfs ability to snapshot for yet another cheap sparse layer of recovery from the unexpected.

> Instead of restarting a chunkserver every night, would be better to add a manual scrub...

Hummm, indeed.  I believe the simplest method for executing this would be for a CS to mount the dataset it is hosting chunks for, which I believe is hard coded to prefer itself for all reads.  Then let it md5sum or find and dd to null the entire self-mounted dataset, thus reading all local chunks which will expose any checksum errors, throw them away, and repair from other presumably valid copies...

**Comment by eriklindahl** [comment]
Oh, I don't think you presume too much, but I do see a danger in presuming the hardware will fail in some ways but not others. Murphy's law tells me hardware is always going to be more creative in finding new unexpected ways to fail than we are in addressing specific failure modes :-)

**Comment by guestisp** [comment]
> Hummm, indeed. I believe the simplest method for executing this would be for a CS to mount the dataset it is hosting chunks for, which I believe is hard coded to prefer itself for all reads. Then let it md5sum or find and dd to null the entire self-mounted dataset, thus reading all local chunks which will expose any checksum errors, throw them away, and repair from other presumably valid copies...

This could be archived, at zero cost, by using ZFS. ZFS automatically detect rotted files.
What I mean with "scrub", in LizardFS worlds, is a cluster-wide scrub: Lizard should scan all files and check if all goals are met, keeping the cluster in a consistent state.

**Comment by 4Dolio** [comment]
Yes. A zfs scrub forces data verification at the zfs layer.  But we are discussing our trust in the LFS layer and ways to force verifying chunk files are valid.  A zfs scrub does not do that as you demonstrated that missing chunk files are not noticed in real time.  The LFS chunk loop is very slow, for good reasons.. so I offered a solution which would force LFS to validate all chunks, while running, by ensuring read attempts of chunk files on a specific CS you want to test.

Such a test (CS self client and md5sum walk) would go further than the CS stop/start re-index of chunks files which only tests for the chunk files existence but not their contents.  My proposed test would be equivalent to a zfs scrub at the LFS data checksum layer.

It would be an interesting test, you could overwrite chunk files with garbage as part of the test.  I do not believe i have encountered this type of fault in production. But if i did it would be transparently handled by LFS...

**Comment by guestisp** [comment]
What about a very very simple:

`find -type f -exec dd if={} of=/dev/null count=1 bs=1M iflag=direct \;` done on mountpoint from the client?

This should read all files, 1MB only, and triggering a crc checksum verify. If reading just 1MB is not enough, this could be removed from the command line.

**Comment by guestisp** [comment]
Anyway, proper scrub feature was asked by me on April: #541

**Comment by 4Dolio** [comment]
Yes. Your find command is exactly what i suggested a few posts ago.  I suggested md5sum instead of dd to null but they accomplish the same goal of reading files..  Since a CS is hard coded to prefer itself (i believe) when acting as a client, you would want to run such a lfs-find-scrub from each chunk server you are attempting to scrub.  Running from a random non CS client would only read one of X chunks from a random CS.  Topology could maybe be used to try and steer such a client, but i suspect mounting from a CS would be more straight forward.

Having used LFS in production for years, chunk bit-rot or underlying weak filesystem errors has never been a problem.

And i have leveraged LFS in some rather strange ways.. the only time I did have problems was when running Xen on raid5 SSDs, and CS VMs using the Xen local lvm as storage, CS disks formatted with ZFS. As a cluster where masters were also clients re-exporting as iet iSCSI, which Xen then mounted for vm store, and then with VMs hosted my that storage... using any fs except zfs on the VM CS worked fine, but for some unknown reason zfs cause chunk files to be improperly read ( even though they seemed completely valid as far as i could deduce )...

**Comment by guestisp** [comment]
What i've proposed:

> What about a very very simple:
>
> `find -type f -exec dd if={} of=/dev/null count=1 bs=1M iflag=direct \; done` on mountpoint from the client?
>
>This should read all files, 1MB only, and triggering a crc checksum verify. If reading just 1MB is not 
> enough, this could be removed from the command line.

won't work, as lizard doesn't read all chunks from all chunkserver. If you have replica 3, a single file read would only read 1 chunk from one of available chunkservers.

I've manually rotted a chunk and tried to see if reading the whole file back would have triggered the crc mismatch, but it doesn't. This because the rotted file was on a chunkserver not used when reading that single chunk.

So, the only realiable way is to implement a scrub feature, that will scrub all chunks (exactly like `HDD_TEST` does but in a much faster way and not only 1 per `HDD_TEST_FREQ` seconds)

**Comment by jkiebzak** [comment]
What happens if you have hundreds of millions of files. `find` would have to crawl all if them. 


Sent from my Verizon, Samsung Galaxy smartphone
-------- Original message --------From: "GUEST.it" <notifications@github.com> Date: 10/8/17  9:20 AM  (GMT-05:00) To: lizardfs/lizardfs <lizardfs@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com> Subject: Re: [lizardfs/lizardfs] How fast should re-replication be initiated? (#589) 
What i've proposed:

What about a very very simple:
find -type f -exec dd if={} of=/dev/null count=1 bs=1M iflag=direct \; done on mountpoint from the client?
This should read all files, 1MB only, and triggering a crc checksum verify. If reading just 1MB is not

enough, this could be removed from the command line.

won't work, as lizard doesn't read all chunks from all chunkserver. If you have replica 3, a single file read would only read 1 chunk from one of available chunkservers.
I've manually rotted a chunk and tried to see if reading the whole file back would have triggered the crc mismatch, but it doesn't. This because the rotted file was on a chunkserver not used when reading that single chunk.
So, the only realiable way is to implement a scrub feature, that will scrub all chunks (exactly like HDD_TEST does but in a much faster way and not only 1 per HDD_TEST_FREQ seconds)

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/lizardfs/lizardfs","title":"lizardfs/lizardfs","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/lizardfs/lizardfs"}},"updates":{"snippets":[{"icon":"PERSON","message":"@guestisp in #589: What i've proposed:\r\n\r\n\u003e What about a very very simple:\r\n\u003e\r\n\u003e `find -type f -exec dd if={} of=/dev/null count=1 bs=1M iflag=direct \\; done` on mountpoint from the client?\r\n\u003e\r\n\u003eThis should read all files, 1MB only, and triggering a crc checksum verify. If reading just 1MB is not \r\n\u003e enough, this could be removed from the command line.\r\n\r\nwon't work, as lizard doesn't read all chunks from all chunkserver. If you have replica 3, a single file read would only read 1 chunk from one of available chunkservers.\r\n\r\nI've manually rotted a chunk and tried to see if reading the whole file back would have triggered the crc mismatch, but it doesn't. This because the rotted file was on a chunkserver not used when reading that single chunk.\r\n\r\nSo, the only realiable way is to implement a scrub feature, that will scrub all chunks (exactly like `HDD_TEST` does but in a much faster way and not only 1 per `HDD_TEST_FREQ` seconds)"}],"action":{"name":"View Issue","url":"https://github.com/lizardfs/lizardfs/issues/589#issuecomment-335006204"}}}

**Comment by 4Dolio** [comment]
> I've manually rotted a chunk and tried to see if reading the whole file back would have triggered the crc mismatch, but it doesn't. This because the rotted file was on a chunkserver not used when reading that single chunk.

Did you run the read test from just any client? Or as a client from the CS you want scrubbed?

**Comment by guestisp** [comment]
Run from the only client I have
I don't run clients on the same chunkservers, chunkserver acts only as chunkservers

**Comment by 4Dolio** [comment]
I understand. but a client on a chunk server will always read from itself. So your scrub should work if you mount your data from the CS you want to test.

**Comment by guestisp** [comment]
As wrote, I don't have any clients on chunkservers
Our client is a stand alone machine

**Comment by 4Dolio** [comment]
apt-get install lizardsf-client # ?

Sigh. if you are trying to do the dd lfs scrub trick it will not work from just any client, so don't say it does not work.  If you will not install the client on your chunk server for the purpose of attempting to scrub-read chunks then you can not run the test you are attempting.

If i want to help my wife set the clock in her car but i refuse to get into her car to do it I can not claim that her car clock can not be set because when i set the clock in my car it does not work.

**Comment by guestisp** [comment]
Is not the same.
I already have a mounted client, why I have to manually scrub all chunkservers when doing the same from a single client should do the scrub?

Is much better to have lizard to check if file being read is consistent across all chunkservers than having to scrub all chunkservers manually

This is what I would expect from a distributed FS

if you scrub a raid array, you scrub the whole array at once, not each single disk manually one by one.

**Comment by 4Dolio** [comment]
Sigh.. that's the thing. You should not need to scrub lfs. And there is no such feature for lfs, yet. The chunk check is for lazy bit rot, but super slow.

So if you insist that you want to lfs scrub then the best option to approximate that, is the procedure(s) we have been arguing about over multiple threads.

You might be able to get closer to the scrub you are attempting, from the client you already have, if you use topology tuning so you client will prefer one specific chunk server. But that won't be as good as doing if from the CS you want to emulate a scrub of.

**Comment by guestisp** [comment]
As wrote, there is a need to scrub because even by reading a file from a client, lizard doesn't check crc for all copies related to the read file.

So if you have some files archived, you'll risk that all reads are made from chunkserver1 and no one is able to know that chunks on chunkserver2 is rotted. When chunkserver1 fails, you loose data as the only survived server has a rotted file

That's why I'm trying to figure out how to scrub all replicas. Even ZFS doesn't protect you against this because if you don't use raid, zfs can do nothing with rotted file and lizard would be unaware of this

I always thought that lizard checksumming on read was made checking all replicas, not only one

**Comment by 4Dolio** [comment]
Lfs only checks the one randomly chosen chunk you happened to read.  Always has, it is wasteful to read them all.

You have been given all the information you need in order to properly replicate a scrub of a given CS.  If you refuse to install the client on a CS to perform that procedure then no one in the world can help you.  So install the client and perform your bit rot scrub test and see if it works.  Or wait for a native scrub feature to be added, if ever it is.

Idk why we keep going back and forth on this.
* such a feature does not exist.
* it is already possible to emulate the desired feature.

**Comment by guestisp** [comment]
@eriklindahl
> We have nice automated failover (and recovery) of the master working with keepalived

Could you please share your keepalive configuration and up/down script?
